runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,728 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+
+   if (!isResolved)
+      {
+      // Resolve and populate dataSnippet fields.
+      generateFillInDataBlockSequenceForUnresolvedField(node, owningMethod, dataSnippet, isWrite, cg);
+      }
+
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *j9classReg = cg->allocateRegister();
+   TR::Register *j9classFlags = cg->allocateRegister();
+ 
+   // Load the class of the instance object into j9classReg.
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, j9classReg, generateS390MemoryReference(objReg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   TR_ASSERT_FATAL(J9ClassHasWatchedFields >= SHRT_MIN  && J9ClassHasWatchedFields <= SHRT_MAX, "expect value of J9ClassHasWatchedFields to be with in 16 bits");
+
+   // Test the bit in fieldClassReg with the J9ClassHasWatchedFields constant to make sure fieldwatch is enabled.
+   auto instr = generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, j9classFlags, generateS390MemoryReference(j9classReg, ((TR_J9VMBase *)(cg->fe()))->getOffsetOfClassFlags(), cg));
+   generateRIInstruction(cg, TR::InstOpCode::NILL, node, j9classFlags, J9ClassHasWatchedFields);
+   cg->setImplicitExceptionPoint(instr);
+   instr->setNeedsGCMap(0x0000FFFF);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+   // If fieldwatch is not enabled, we can branch to cFlowRegionEnd directly.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BZ, node, cFlowRegionEnd);
+
+   uint8_t numOfConditions = getNumOfConditionsForReportFieldAccess(node, isResolved, isWrite, true /*isInstanceField*/, cg);
+
+   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions(0, numOfConditions, cg);
+   deps->addPostCondition(j9classReg, TR::RealRegister::AssignAny);
+
+   // Now we finally generate instructions to report the fieldwatch to the appropriate VM helper routine.
+   generateReportFieldAccessOutlinedInstructions(node, cFlowRegionEnd, dataSnippet, isWrite, deps, cg);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionEnd, deps);
+
+   cg->stopUsingRegister(j9classFlags);
+   cg->stopUsingRegister(j9classReg);
+   }
+
+void J9::Z::TreeEvaluator::VMrdbarEvaluatorForFieldWatch(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   bool isStatic = node->getSymbolReference()->getSymbol()->getKind() == TR::Symbol::IsStatic;
+   J9Method *owningMethod = (J9Method *)TR::TreeEvaluator::getMethodFromBCInfo(node->getByteCodeInfo(), cg->comp());
+   int32_t bcIndex = node->getByteCodeInfo().getByteCodeIndex();
```

Comment:
It was just to avoid doing the same work in both of the routines. And now that all codgens will share this routine, we won't have to do this work in all of the helpers. If you prefer to still offload this work to the `*WriteBarrierHelperForFieldWatch` helpers then I'll move it there.

Reply:
I think we can refactor this and `VMrdbarEvaluatorForFieldWatch` into the same routine and move it to the common codegen as well.

Reply:
424b8ba